packages/connection/src/class-nonce-handler.php

```diff
@@ -0,0 +1,197 @@
+<?php
+/**
+ * The nonce handler.
+ *
+ * @package automattic/jetpack-connection
+ */
+
+namespace Automattic\Jetpack\Connection;
+
+/**
+ * The nonce handler.
+ */
+class Nonce_Handler {
+
+	/**
+	 * How many nonces should be removed during each run of the runtime cleanup.
+	 * Can be modified using the filter `jetpack_connection_nonce_cleanup_runtime_limit`.
+	 */
+	const CLEANUP_RUNTIME_LIMIT = 10;
+
+	/**
+	 * How many nonces should be removed per batch during the `clean_all()` run.
+	 */
+	const CLEAN_ALL_LIMIT_PER_BATCH = 1000;
+
+	/**
+	 * Nonce lifetime in seconds.
+	 */
+	const LIFETIME = HOUR_IN_SECONDS;
+
+	/**
+	 * The nonces used during the request are stored here to keep them valid.
+	 *
+	 * @var array
+	 */
+	private static $nonces_used_this_request = array();
+
+	/**
+	 * Adds a used nonce to a list of known nonces.
+	 *
+	 * @param int    $timestamp the current request timestamp.
+	 * @param string $nonce the nonce value.
+	 * @param bool   $run_cleanup Whether to run the `cleanup_runtime()`.
+	 *
+	 * @return bool whether the nonce is unique or not.
+	 */
+	public static function add( $timestamp, $nonce, $run_cleanup = true ) {
+		global $wpdb;
+
+		if ( isset( static::$nonces_used_this_request[ "$timestamp:$nonce" ] ) ) {
+			return static::$nonces_used_this_request[ "$timestamp:$nonce" ];
+		}
+
+		// This should always have gone through Jetpack_Signature::sign_request() first to check $timestamp an $nonce.
+		$timestamp = (int) $timestamp;
+		$nonce     = esc_sql( $nonce );
+
+		// Raw query so we can avoid races: add_option will also update.
+		$show_errors = $wpdb->show_errors( false );
+
+		$old_nonce = $wpdb->get_row(
+			$wpdb->prepare( "SELECT * FROM `$wpdb->options` WHERE option_name = %s", "jetpack_nonce_{$timestamp}_{$nonce}" )
+		);
+
+		if ( is_null( $old_nonce ) ) {
+			$return = (bool) $wpdb->query(
+				$wpdb->prepare(
+					"INSERT INTO `$wpdb->options` (`option_name`, `option_value`, `autoload`) VALUES (%s, %s, %s)",
+					"jetpack_nonce_{$timestamp}_{$nonce}",
+					time(),
+					'no'
+				)
+			);
+
+			/**
+			 * Use the filter to disable the nonce cleanup that happens at shutdown after adding a new nonce.
+			 *
+			 * @since 9.0.0
+			 *
+			 * @param int $limit How many old nonces to remove at shutdown.
+			 */
+			if ( apply_filters( 'jetpack_connection_add_nonce_cleanup', $run_cleanup ) ) {
+				add_action( 'shutdown', array( __CLASS__, 'clean_runtime' ) );
+			}
+		} else {
+			$return = false;
+		}
+
+		$wpdb->show_errors( $show_errors );
```

Comment:
Do we care about the possibility of anything between line 59 and here throwing, and so preventing the show_errors from being reset?

Reply:
Good catch, thanks for forcing me to dig this up :)

This piece of history was added 9 years ago and released in `Jetpack 1.2.3`, fascinating stuff: https://github.com/Automattic/jetpack/commit/51fe166af6a968db3c65b38aa56b8e41c903d439

Since then it traveled all over Jetpack:
- `jetpack.php`
- `class.jetpack.php`
- `class-manager.php`
- and now we're moving it into `class-nonce-handler.php`

I'm guessing it was added to gracefully catch the "duplicate entry" error before we added the `SELECT` there, so it was kind of two requests in one: `INSERT` checked if the entry exists, and added it if it didn't.

On the other hand, it seems like hiding errors is still relevant today in case of a race condition. If two requests with the same nonce come simultaneously, that nonce may be added between the `SELECT` and `INSERT`, making `INSERT` throw the "duplicate entry" error.

So I would say we'd better leave it there.

Reply:
I think you answered a different question than what I was asking. But that was interesting, thanks for the explanation!

I was just wondering whether we should have a `try { ... } finally { $wpdb->show_errors( $show_errors ); }` or something like that, or if we expect that any exception thrown here is going to be uncaught or caught by a fatal error handler of some sort so the show_errors flag remaining false wouldn't be a problem.

Reply:
As for the possible race leading to a "duplicate entry" error, there's a simple enough solution for that: use `INSERT IGNORE`, which will have MySQL specifically avoid raising the error for the INSERT. Some quick testing suggests that `$wpdb->query()` will return 0 for an ignored insert, versus 1 for a successful insert and false for a duplicate entry without IGNORE.

You could also use a transaction, and `SELECT ... FOR UPDATE` in the original select. But MySQL can get a bit funny about the gap locks involved when using `SELECT ... FOR UPDATE` to guard an INSERT: if two processes both lock the same gap with `SELECT ... FOR UPDATE`, then their INSERTs will block each other even if the inserts wouldn't actually conflict. Sigh.

Reply:
Good idea, I added that 👍